[FIX] web: mitigate navigation error on iOS/iPadOS devices

detrouxdev opened the pull request:
Ever since iOS 16.4, going back and forth through a website with the cache enabled, creates an error in the console which Odoo tries to handle but fails to do so. Therefor a generic CORS error message appears.

There is unfortunately no proper way to fix this bug on our end so instead, this commit tries to mitigate the error by returning early and not showing the traceback dialog.

I (ARD) have submitted a feedback through Apple's Feedback Assistant app and will subsequently remove this commit when Apple releases an update that fixes this.

opw-3281727

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/124828).

rdeodoo commented:
Mobile and js framework opinion asked here https://discord.com/channels/678381219515465750/687338039717920792/1118172385078685766

detrouxdev commented:
@rdeodoo I made the following changes:
- Remove the version check, instead I just check if it's safari
- I added the check if it's in the frontend
- I also added a check if we're not in debug=assets so that we can easily check any error by enabling debug assets mode.

Let me know what you think

pparidans commented on addons/web/static/src/js/services/crash_manager.js:
```diff
@@ -122,6 +123,19 @@ var CrashManager = AbstractService.extend({
+                    // cache is enabled. A feedback has been reported but in the
+                    // meantime, hide any script error in these versions.
+                    if (
+                        /^((?!chrome|android).)*safari/i.test(navigator.userAgent)
+                        && ("maxTouchPoints" in navigator)
```
:warning: Don't forget that *every browser* on iOS uses WKWebView as rendering engine, thus all are impacted in some way by this issue (tested at least with Firefox for iOS). So, I wouldn't filter too much...

BTW, `web.config` already implements a `isIOS` flag for that purpose.

rdeodoo commented on addons/web/static/src/js/services/crash_manager.js:
```diff
@@ -122,6 +124,18 @@ var CrashManager = AbstractService.extend({
+                    // meantime, hide any script error in these versions.
+                    if (
+                        config.device.isIOS
+                        && message === "Script error."
```
Seems useless as the famous `Script error.` means that we are in the CORS error case that we are catching just a few lines above through `if (!file && !line && !col) {` in the Chrome/Opera/Safari case.
So it's kinda redundant, but it's fine.

https://github.com/odoo/odoo/pull/21032#issuecomment-367753861

> Browser differences:
>   
> 1. Chrome & Opera: `message = 'Script Error.'`, `error = null`
> 2. Firefox [is handling it directly](https://dxr.mozilla.org/mozilla-beta/source/dom/base/nsJSEnvironment.cpp#464) and return `uncaught exception: [object Object]` instead of `Script Error.` for `message` and set `error` to the error with the message object that Chrome & Opera are blocking.
> 3. Safari: Need to check, but since it is a webkit browser, it most likely handle it like Chrome & Opera 
> 
> [Mozilla 'Script Error.' handler](https://dxr.mozilla.org/mozilla-beta/source/dom/base/nsJSEnvironment.cpp#464)
[WebKit 'Script Error.'](http://trac.webkit.org/browser/branches/chromium/648/Source/WebCore/dom/ScriptExecutionContext.cpp?rev=77122#L301)
> Coming from this [stackoverflow](https://stackoverflow.com/questions/5913978/cryptic-script-error-reported-in-javascript-in-chrome-and-firefox)
> 
> Found this too:
https://bugs.webkit.org/show_bug.cgi?id=132945
https://bugs.webkit.org/show_bug.cgi?id=70574

rdeodoo reviewed (approved):
@robodoo r+

Ty for the fix, good idea :+1: 

Tested and working fine. Since we target only the frontend, we shouldn't impact any flows, as I [introduced](https://github.com/odoo/odoo/pull/21032) that CORS error catching for google analytics in the backend.
Note that for google map configuration in frontend (edit mode): it is doing calls to their server and properly able to catch and display errors, as opposed to the backend analytics which couldn't listen the error as it comes from the injected google script.
